pkg/remote/crd_upgrade.go

```diff
@@ -21,6 +29,142 @@ func PatchCRD(ctx context.Context, clnt client.Client, crd *v1extensions.CustomR
 		client.FieldOwner(v1beta2.OperatorName))
 }
 
+type CrdType string
+
+const (
+	KCP CrdType = "KCP"
+	SKR CrdType = "SKR"
+)
+
+func updateRemoteCRD(
+	ctx context.Context, plural string, kyma *v1beta2.Kyma, runtimeClient Client, controlPlaneClient Client) (
+	bool, *v1extensions.CustomResourceDefinition, *v1extensions.CustomResourceDefinition, error) {
+	crd := &v1extensions.CustomResourceDefinition{}
+	crdFromRuntime := &v1extensions.CustomResourceDefinition{}
+	var err error
+	err = controlPlaneClient.Get(
+		ctx, client.ObjectKey{
+			// this object name is derived from the plural and is the default kustomize value for crd namings, if the CRD
+			// name changes, this also has to be adjusted here. We can think of making this configurable later
+			Name: fmt.Sprintf("%s.%s", plural, v1beta2.GroupVersion.Group),
+		}, crd,
+	)
+
+	if err != nil {
+		return false, nil, nil, err
+	}
+
+	err = runtimeClient.Get(
+		ctx, client.ObjectKey{
+			Name: fmt.Sprintf("%s.%s", plural, v1beta2.GroupVersion.Group),
+		}, crdFromRuntime,
+	)
+
+	if ShouldPatchRemoteCRD(crdFromRuntime, crd, kyma, err) {
+		err = PatchCRD(ctx, runtimeClient, crd)
+		if err != nil {
+			return false, nil, nil, err
+		}
+
+		err = runtimeClient.Get(
+			ctx, client.ObjectKey{
+				Name: fmt.Sprintf("%s.%s", plural, v1beta2.GroupVersion.Group),
+			}, crdFromRuntime,
+		)
+		if err != nil {
+			return false, nil, nil, err
+		}
+
+		return true, crdFromRuntime, crd, nil
+	}
+
+	if err != nil {
+		return false, nil, nil, err
+	}
+
+	return false, nil, nil, nil
+}
+
+func ShouldPatchRemoteCRD(
+	runtimeCrd *v1extensions.CustomResourceDefinition, kcpCrd *v1extensions.CustomResourceDefinition,
+	kyma *v1beta2.Kyma, err error) bool {
+	if k8serrors.IsNotFound(err) {
+		return true
+	}
+	kcpAnnotation, _ := getAnnotation(kcpCrd, KCP)
+	skrAnnotation, _ := getAnnotation(runtimeCrd, SKR)
+
+	latestGeneration := strconv.FormatInt(kcpCrd.Generation, 10)
+	runtimeCRDGeneration := strconv.FormatInt(runtimeCrd.Generation, 10)
+	return !containsLatestCRDGeneration(kyma.Annotations[kcpAnnotation], latestGeneration) ||
+		!containsLatestCRDGeneration(kyma.Annotations[skrAnnotation], runtimeCRDGeneration)
+}
+
+func updateKymaAnnotations(kyma *v1beta2.Kyma, crd *v1extensions.CustomResourceDefinition, crdType CrdType) error {
+	if kyma.Annotations == nil {
+		kyma.Annotations = make(map[string]string)
+	}
+	annotation, err := getAnnotation(crd, crdType)
+	if err != nil {
+		return err
+	}
+	kyma.Annotations[annotation] = strconv.FormatInt(crd.Generation, 10)
+	return nil
+}
+
+func getAnnotation(crd *v1extensions.CustomResourceDefinition, crdType CrdType) (string, error) {
```

Comment:
Actually, you can create the annotation dynamically and don't need to return the error, something like:
```
func getAnnotation(crd *v1extensions.CustomResourceDefinition, crdType CrdType) (string) {
     return fmt.sprintf("%s-%s-crd-generation", crd.Spec.Names.Kind, crdType)
}
```

And since you have this function, don't need to declare a cons value in the `operator_annotations.go`